File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        std::stack<std::pair<int, int>> stk;
        int max_area = 0;
        int n = heights.size();
        for (int i = 0; i < n; ++i) {
            int start = i;
            while (!stk.empty() && stk.top().second > heights[i]) {
                std::pair<int, int> top = stk.top(); stk.pop();
                int index = top.first;
                int h = top.second;
                max_area = max(max_area, h * (i - index));
                start = index;
            }
            stk.push({start, heights[i]});
        }
        while (!stk.empty()) {
            pair<int, int> top = stk.top(); stk.pop();
            max_area = max(max_area, top.second * (n - top.first));
        }
        return max_area;
    }
};

File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        stack = []
        max_area = 0
        for i, h in enumerate(heights):
            start = i
            while stack and stack[-1][1] > h:
                idx, hei = stack.pop()
                max_area = max(max_area, hei * (i - idx))
                start = idx
            stack.append([start, h])
        for i, h in stack:
            max_area = max(max_area, (len(heights) - i) * h)
        return max_area